File: scripts/mini-bench-setup.py
#!/usr/bin/env python3
"""
mini-bench-setup.py - One-time corpus indexing for mini-bench.
Run once; the DB is reused on every bench run.
"""

import os
import subprocess
import sys
import time

BENCH_DB = "/tmp/quaid-mini-bench.db"
CORPUS_DIR = "/tmp/quaid-bench-corpus"
QUAID = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                     "target", "release", "quaid")

env = {**os.environ, "QUAID_DB": BENCH_DB}


def run(args, **kwargs):
    return subprocess.run([QUAID] + args, env=env, capture_output=True, text=True, **kwargs)


def main():
    if not os.path.isdir(CORPUS_DIR):
        print(f"ERROR: corpus not found at {CORPUS_DIR}")
        print("Run the DAB corpus generator first:")
        print("  cd ~/repos/quaid-evals && python3 benchmarks/dab/generate_corpus.py")
        sys.exit(1)

    if os.path.exists(BENCH_DB):
        os.unlink(BENCH_DB)

    print(f"Initialising DB at {BENCH_DB}...")
    r = run(["init", BENCH_DB])
    if r.returncode != 0:
        print(f"init failed: {r.stderr}")
        sys.exit(1)

    print(f"Adding corpus collection from {CORPUS_DIR}...")
    r = run(["collection", "add", "docs", CORPUS_DIR])
    if r.returncode != 0:
        print(f"collection add failed: {r.stderr}")
        sys.exit(1)

    print("Syncing collection...")
    r = run(["collection", "sync", "docs"])
    if r.returncode != 0:
        print(f"sync failed: {r.stderr}")
        sys.exit(1)

    print("Generating embeddings (this takes ~60-120s on first run)...")
    t0 = time.time()
    r = run(["embed", "--all"], timeout=300)
    elapsed = time.time() - t0
    if r.returncode != 0:
        print(f"embed failed: {r.stderr}")
        sys.exit(1)

    print(f"Done in {elapsed:.0f}s. Corpus indexed and ready at {BENCH_DB}")


if __name__ == "__main__":
    main()

File: scripts/mini-bench-setup.mjs
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { CORPUS_DIR, generateCorpus } from "./mini-bench-corpus.mjs";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPO_DIR = path.dirname(__dirname);
const DEFAULT_DB = "/tmp/quaid-mini-bench.db";
const DEFAULT_QUAID = path.join(REPO_DIR, "target", "release", "quaid");

function parseArgs(argv) {
  const args = {
    db: DEFAULT_DB,
    quaid: DEFAULT_QUAID,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--db") {
      args.db = argv[++i];
    } else if (arg === "--quaid") {
      args.quaid = argv[++i];
    } else {
      throw new Error(`unknown argument: ${arg}`);
    }
  }

  return args;
}

function run(quaid, args, options = {}) {
  const result = spawnSync(quaid, args, {
    cwd: REPO_DIR,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    ...options,
  });

  return result;
}

function fail(step, result) {
  const detail = [result.stdout, result.stderr].filter(Boolean).join("\n").trim();
  console.error(`${step} failed${detail ? `:\n${detail}` : ""}`);
  process.exit(result.status || 1);
}

const args = parseArgs(process.argv.slice(2));
const quaid = path.resolve(args.quaid);

if (!fs.existsSync(quaid)) {
  console.error(`ERROR: quaid binary not found at ${quaid}`);
  console.error("Run: cargo build --release");
  process.exit(1);
}

console.log(`Generating corpus at ${CORPUS_DIR}...`);
generateCorpus(CORPUS_DIR);

if (!fs.existsSync(args.db)) {
  console.log(`Initialising DB at ${args.db}...`);
  const init = run(quaid, ["init", args.db]);
  if (init.status !== 0) fail("init", init);
} else {
  console.log(`Using existing DB at ${args.db}...`);
}

console.log(`Adding corpus collection from ${CORPUS_DIR}...`);
const add = run(quaid, ["--db", args.db, "collection", "add", "docs", CORPUS_DIR]);
if (add.status !== 0) {
  const output = `${add.stdout}\n${add.stderr}`;
  if (!/already exists|collection exists|duplicate/i.test(output)) {
    fail("collection add", add);
  }
  console.log("Collection docs already exists; continuing.");
}

console.log("Syncing collection...");
const sync = run(quaid, ["--db", args.db, "collection", "sync", "docs"]);
if (sync.status !== 0) fail("collection sync", sync);

console.log("Generating embeddings...");
const embed = run(quaid, ["--db", args.db, "embed", "--all"], { timeout: 300_000 });
if (embed.status !== 0) fail("embed", embed);

console.log("Ready.");
